src/_posts/2020-04-25-functional-programming-in-elixir.md

```diff
@@ -0,0 +1,281 @@
+    Map.update(acc, word, 1, &(&1 + 1))
+  end
+end
+```
+
+This is easier to understand.
+There is a convention to align the pipeline operator with the data that it was passed into.
+
+We can go much further with this:
+
+### Second Refactor
+
+```
+defmodule Words do
+
+  @seperators ~r/[ _,!&@$%^&:]/u
+
+  def count(sentence) do
+    sentence
+    |> String.downcase()
+    |> String.split(@seperators, trim: true)
+    |> Enum.reduce( %{}, &update_map/2)
+  end
+
+	defp update_map(word, acc) do
+  	Map.update(acc, word, 1, &(&1 + 1))
+  end
+end
+```
+
+This is a clearer read. You now can see the order of execution.
+
+### Third Refactor
+
+Now I am going to add a typespec to the code. The code so far looks untyped, but that is only because we have not restricted anything.
+
+```
+defmodule Words do
+
+  @seperators ~r/[ _,!&@$%^&:]/u
+
+  @spec count(String.t()) :: map
+  def count(sentence) when is_binary(sentence) do
+    sentence
+    |> String.downcase()
+    |> String.split(@seperators, trim: true)
+    |> Enum.reduce( %{}, &update_map/2)
+  end
+
+  @spec counter(list, map) :: map
+	defp update_map(word, acc) do
+  	Map.update(acc, word, 1, &(&1 + 1))
+  end
+end
+```
+
+Here i have made three edits to the solution. Each of the functions now has a `typespec` and I have added a guard clause to `count/1`. Typespecs allow static checking of the code. It's not a required part of the language but does add value in larger projects. There is a tool called [dialyzer](https://hexdocs.pm/dialyzex/Mix.Tasks.Dialyzer.html) that can be used to staticly check a codebase to ensure that all uses of a function conform to the typespec. Recent versions of Elixir (from version 1.10) will check that a functions signature matches the typespec if a typespec is provided.
+
+`is_binary/1` is a guard clause. This can be used to assist with the definition of a type. Functions in Elixir use pattern matching which permits a function to have multiple clauses. Guard clauses provide the ability to add some extra details. The name binary comes from Erlang, since it can be used to parse a binary file.
+
+It's fairly common to have functions return two tuples `{:ok, details}` and `{:error, reason}`.
+Here is a process/1 function:
+
+```
+def process({:ok, details} = body) where is_string(details) do
+  IO.puts(details)
+  body
+end
+
+def process(body = {:error, reason}) do
+  IO.puts("Error, #{reason}")
+  body
+end
+```
+
+This demonstrates a two clause function. If details is not a string then the first clause will not match and you will get a run time exception. The `= body` is used for pattern matching to capture the entire tuple. This matching can happen on the left or the right. Pattern matching is also used to deconstruct the tuple to obtain the details. In both cases the input is passed on to the output making them easy to chain with pipelines. This allows some sophisticated validation to be applied without needing to use an `if`.
```

Comment:
Change to: 

In both cases, the input is passed on to the output making them easy to chain with pipelines.